install: clean up script in `roachprod start`

tbg opened the pull request:
See individual commits.

cc @cockroachdb/test-eng

Release note: None

tbg committed:
install: use go:embed for systemd start script template

This intentionally does not touch the contents of the script,
i.e. does not try to turn it into a vanilla bash script.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/67363)
<!-- Reviewable:end -->

tbg committed:
install: reduce layers of recursion in cockroach start script

Previously, we'd execute a giant string on each node, which

- would write most of itself into a file,
- execute that file, which would in turn
- set up a systemd job that would
- execute that file again (with a different arg),
- which would start CRDB.

This was all better than what we had before, but it wasn't maintainable
by any stretch of imagination. Accepting that we're going to be stuck
with a little bit of bash script in the foreseeable future, this commit
at least cleans things up to the point of being palatable.

Concretely:

- proper escaping for everything that is injected from Go into the
  script. This required some changes (such as not pasting in $HOME
  any more as this would no longer get expanded).
- move all template code to the beginning of the file.
- make the template code look like comments, so that we can
- make the script pass shellcheck.

Also,

- add a datadriven Go unit test that executes the template.
- add an "integration" test, i.e. a little script that actually
  starts a local and remote cluster. This is intentionally not
  part of CI but it's going to be helpful for local development.

I think this should be our Gold standard going forward for what
little shell script we need to maintain. I'm happy to incorporate
additional suggestions on how the bar can be raised, too, but also
don't want to get to the point where migrating off of the shell
scripting requires a bigger lift.

Release note: None

rickystewart reviewed (dismissed):
Can you clean up the commit history?

tbg commented:
Sorry, not ready for review yet. I posted the original commit but then went
berserk. I'll ping when it's ready.

On Fri, Jul 9, 2021 at 4:50 PM Ricky Stewart ***@***.***>
wrote:

> ***@***.**** requested changes on this pull request.
>
> Can you clean up the commit history?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/67363#pullrequestreview-703129838>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABGXPZG6YVSGHGLBU5X5NWTTW4EDHANCNFSM5AASVAZA>
> .
>

tbg commented:
Ready for a look!

tbg commented:
Friendly ping.

stevendanna reviewed (commented):
Left a few comments but nothing too major.  Thanks for tackling this

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67363#-:-MeV9IWl4-kREOFb2jNt:b-632r1z)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/cmd/roachprod/install/cockroach.go, line 249 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67363#-MeV2fR27RRk-6fIK7bP:-MeV2fR27RRk-6fIK7bQ:b-jlkebf) ([raw file](https://github.com/cockroachdb/cockroach/blob/115eb31d1707c82446de437de25f4d37c32dac80/pkg/cmd/roachprod/install/cockroach.go#L249)):*
> ```Go
> // LogDir implements the ClusterImpl.NodeDir interface.
> func (Cockroach) LogDir(c *SyncedCluster, index int) string {
> 	dir := "${HOME}/logs"
> ```

This changes the semantics of this function a little big, but I checked all the callsites and I think this is OK.
___
*[pkg/cmd/roachprod/install/cockroach.go, line 458 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67363#-MeV6TFB-m3g-LFmOvFZ:-MeV6TFB-m3g-LFmOvF_:b-83sjo4) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfec1164fed6b8188f2975a48c084d95de5b5f0/pkg/cmd/roachprod/install/cockroach.go#L458)):*
> ```Go
> 			return shellescape.Quote(fmt.Sprint(i))
> 		}}).
> 		Delims("#{", "#}").
> ```

Nice.
___
*[pkg/cmd/roachprod/install/scripts/start.sh, line 39 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67363#-MeUy14u9qkHVu7wOG2o:-MeUy14u9qkHVu7wOG2p:b6s4b5w) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfec1164fed6b8188f2975a48c084d95de5b5f0/pkg/cmd/roachprod/install/scripts/start.sh#L39)):*
> ```Shell
> # End of templated code.
> 
> ENV_VARS+="ROACHPROD=${NODE_NUM}${TAG}"
> ```

Shellcheck warns on this line with: 

```
In pkg/cmd/roachprod/install/scripts/start.sh line 39:
ENV_VARS+="ROACHPROD=${NODE_NUM}${TAG}"
^------^ SC2179: Use array+=("item") to append items to an array.

For more information:
  https://www.shellcheck.net/wiki/SC2179 -- Use array+=("item") to append ite...
```

https://github.com/koalaman/shellcheck/wiki/SC2179

If the warning is erroneous, might be good to add a comment.
___
*[pkg/cmd/roachprod/install/scripts/start.sh, line 51 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67363#-MeV7Unu5l_SA7V2BiA4:-MeV7Unu5l_SA7V2BiA5:bnlcuz4) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfec1164fed6b8188f2975a48c084d95de5b5f0/pkg/cmd/roachprod/install/scripts/start.sh#L51)):*
> ```Shell
>   mkdir -p "${LOG_DIR}"
>   echo "cockroach start: $(date), logging to ${LOG_DIR}" | tee -a "${LOG_DIR}"/{roachprod,cockroach.std{out,err}}.log
>   if [[ -n "${KEY_CMD}" ]]; then
> ```

I suppose we don't technically need this check since eval will happily eval an empty string, but it doesn't hurt.
___
*[pkg/cmd/roachprod/install/scripts/start.sh, line 54 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67363#-MeV7nyTBOHHgoXMRncU:-MeV7nyTBOHHgoXMRncV:b-t8vzfl) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfec1164fed6b8188f2975a48c084d95de5b5f0/pkg/cmd/roachprod/install/scripts/start.sh#L54)):*
> ```Shell
>     eval "${KEY_CMD}"
>   fi
>   export "${ENV_VARS[@]}"
> ```

note: if ENV_VARS is empty, this will end up printing the currently exported environment. 


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> Dismissed @stevendanna from 3 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67363)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @stevendanna)

___
*[pkg/cmd/roachprod/install/scripts/start.sh, line 39 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67363#-MeUy14u9qkHVu7wOG2o:-MfY-pcK489me-x4zNTg:b-rgux6t) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfec1164fed6b8188f2975a48c084d95de5b5f0/pkg/cmd/roachprod/install/scripts/start.sh#L39)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Shellcheck warns on this line with: 

```
In pkg/cmd/roachprod/install/scripts/start.sh line 39:
ENV_VARS+="ROACHPROD=${NODE_NUM}${TAG}"
^------^ SC2179: Use array+=("item") to append items to an array.

For more information:
  https://www.shellcheck.net/wiki/SC2179 -- Use array+=("item") to append ite...
```

https://github.com/koalaman/shellcheck/wiki/SC2179

If the warning is erroneous, might be good to add a comment.
</blockquote></details>

Seems legit, fixed.
___
*[pkg/cmd/roachprod/install/scripts/start.sh, line 54 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67363#-MeV7nyTBOHHgoXMRncU:-MfY0Ncf07qw2a_UfSWc:b-s4cciz) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfec1164fed6b8188f2975a48c084d95de5b5f0/pkg/cmd/roachprod/install/scripts/start.sh#L54)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

note: if ENV_VARS is empty, this will end up printing the currently exported environment. 
</blockquote></details>

Added a comment, but luckily it's never empty.


<!-- Sent from Reviewable.io -->

tbg commented:
bors r=stevendanna

RaduBerinde commented:
___
*[pkg/cmd/roachprod/install/scripts/start.sh, line 41 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67363#-MooGA6l7FBl1-TIeugd:-MooGA6l7FBl1-TIeuge:b-8lwfpw) ([raw file](https://github.com/cockroachdb/cockroach/blob/0de4773f4ed9d83d52c05200b0ae212152cd3551/pkg/cmd/roachprod/install/scripts/start.sh#L41)):*
> ```Shell
> ENV_VARS+=("ROACHPROD=${NODE_NUM}${TAG}")
> if [[ -n "${ADVERTISE_FIRST_IP}" ]]; then
>   ARGS+=("--advertise-host" "$(hostname -I | awk '{print $1}')")
> ```

Does anyone know why we do this dance instead of using `cloud.VM.PublicIP` from the roachprod code? Using `PublicIP` should work (and be more robust) no?


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/cmd/roachprod/install/scripts/start.sh, line 41 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/67363#-MooGA6l7FBl1-TIeugd:-MooOPLg1D3pSKXMDDTC:b-xy7oz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/0de4773f4ed9d83d52c05200b0ae212152cd3551/pkg/cmd/roachprod/install/scripts/start.sh#L41)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Does anyone know why we do this dance instead of using `cloud.VM.PublicIP` from the roachprod code? Using `PublicIP` should work (and be more robust) no?
</blockquote></details>

Err, I think I mean `PrivateIP`


<!-- Sent from Reviewable.io -->

tbg commented:
I don't know, I say try if it works. Good chance this code is just old.

(sent from mobile, please excuse my brevity)

On Thu, Nov 18, 2021, 19:53 RaduBerinde ***@***.***> wrote:

> ------------------------------
>
> *pkg/cmd/roachprod/install/scripts/start.sh, line 41 at r5
> <https://reviewable.io/reviews/cockroachdb/cockroach/67363#-MooGA6l7FBl1-TIeugd:-MooOPLg1D3pSKXMDDTC:b-xy7oz9>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/0de4773f4ed9d83d52c05200b0ae212152cd3551/pkg/cmd/roachprod/install/scripts/start.sh#L41>):*
> *Previously, RaduBerinde wrote…*
>
> Does anyone know why we do this dance instead of using cloud.VM.PublicIP
> from the roachprod code? Using PublicIP should work (and be more robust)
> no?
>
> Err, I think I mean PrivateIP
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/67363#issuecomment-973161306>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABGXPZHWXNJHGCS4MCNLSDDUMVDUNANCNFSM5AASVAZA>
> .
>
